fix redirects (#105) and mock nesting

clslgrnc opened the pull request:
I took the example from #105 which does indeed raise an error on `master`
~To fix the problem I took the same approach used for send: when `NoMockAddress` is raised while `self._real_http` is `True` or when `_RunRealHTTP` is raised I use `_original_get_adapter`.~

**Edit**: I extended the test to include mock nesting (since my first commit completely bypasses any outer mock).
To fix the problem and improve mock nesting:
- `requests.Session.get_adapter` is restored from a value taken at the start of the current mock, instead of the beggining of the `_fake_send` function (this allows redirects)
- `_last_send` is used when `NoMockAddress` is raised while `self._real_http` is `True`, or when `_RunRealHTTP` is raised, in order to pass the request to the outer mock

After this fix the added test passes without error.

~(I am not sure `_last_send` or `_last_get_adapter` offer any useful functionality though)~

clslgrnc commented:
Codacy complains about constants that do not conform to UPPER_CASE naming style:
- `_original_get_adapter` was chosen for consistency with `_original_send`;
- `HTTPStatus_FOUND` was chosen for its proximity to `HTTPStatus.FOUND` (the introduction of this variable is a workaround to the fact that python2.7 does not have `HTTPStatus`.

But I do not have any objection to changing any to upper case.

clslgrnc commented:
Since it seems the semantic of nested mocks is not entirely clear. I'll explain here how I understand it in this PR (and hopefully how it behaves in this PR).

The way I understand it, `Mocker` creates a "fake internet": any call to `requests.Session.send` is intercepted and the chosen fake response corresponding to the input address is returned.
If no response is configured for a given address, it either:
- raises `NoMockAddress`
- return the response from the "real internet" if `_real_http` is set to True (globally for the Mocker or specifically for the given address)

It follows naturally that with nested mocks, the "fake internet" created by a mock is the "real internet" of the mocks it immediately contains.

joshvillbrandt commented:
Woohoo! This fixed my nesting issues. 🎉

Please merge this and release an update to requests-mock on PyPi. Thanks, fam! 👊

jamielennox commented:
heh, sorry i've been moving around a lot and changing jobs. I promise to have a look, just need to be careful that this isn't a behaviour change for anyone. 

Also it's honestly a section of the code that i haven't touched much since i initially wrote it, so i need to wrap my head around it again.

clslgrnc commented:
No worries.
I guess it *will* change the behaviour of nesting since the added test did not pass before the patch, it did not break any other tests though.
I do not know how much this behaviour change will impact the behaviour people relied on (actually I am not even sure what the behaviour was before).

One potential change I haven't checked is that `real_http` now jumps to the mock immediately containing the current mock. Maybe the previous behavior was to completely bypass all mocks (I am not sure).
If this is the case, or a desired feature, we may need two different option:
- one to leave the current mock
- one to leave all mocks

joshvillbrandt commented:
Bump! :)

clslgrnc commented:
Concerning behavior change, since there is no documentation on nesting (correct me if I am wrong), I looked at issues mentioning nesting and found: #27, #44, #82, #105. Let me know if there are other relevant issues.

The first one, #27, seems to show that there was a behavior change between 0.6 and 0.7 (in 0.7 inner mock simply replaces the outter mock) and that at least some people expect mocks to be nestable in particular "I think the ideal behavior would be to try the inner mock first, and if there's no match go to the outer mock" (David Chudzicki).

The behavior I propose is (almost) consistent with both:
- if `real_http` is set to `False` at the mock level, the inner mock replaces the outter mock,
- if `real_http` is set to `True` the inner mock is searched first, then the outter mock is used.

There is a difference though if `real_http` is set to `True` at the resource level in the inner mock then it will search the outter mock instead of jumping directly to an unmocked request (if that is what was happening).

What I propose is also consistent with your suggestion of having a `new` method (without the need of a `new` method):
```python
with self.requests_mock.new() as mocker:
   mocker.get('http://test.com', json={})
```

On #44 I agree with your suggestion to avoid nesting in this case, but this seem to be how some users use requests-mock. I believe this PR would also have solved that user's issue (provided he set `real_http` to `True`).

#82 does not seem relevant and nesting is only mentioned in passing in #105. Still, #105 is relevant because this PR fixes #105:
- if you take just the first commit it forces inner mocks to bypass any outter mock on `real_http`
- if you take the whole PR it pass the current request to the immediate outter mock on `real_http`

I prefer the second one because:
- requests_mock `real_http` itself can be tested without relying on the _real_ internet
- users can define a mock at the test class level that handle all authentication code and block access to the real internet, while test writers can include request mocks without having to know about this outter mock

If you want I can update the documentation to add a section on nesting.

joshvillbrandt commented:
The latest commit continues to work for my mock nesting needs. 👍

Seems like you've done an excellent job reviewing the needs of the community and making sure everything has been addressed. Thanks for your work, @clslgrnc!

clslgrnc commented:
You're welcome, at least it fulfills your (and my) nesting needs (and fixes redirects).
The last force-push was just a rebase and a slight simplification of the PR.

clslgrnc commented:
With COVID-19 and his new job, @jamielennox probably has other priorities right now (I hope you're ok). If anyone needs nested mocks, as a temporary workaround you can use:
```
pip install git+git://github.com/clslgrnc/requests-mock@6d36219b1cc08111ddbf78ab5df0732874c5812e
```
:warning: **Do not blindly pull code from random people** into your development or continuous integration (or any) environment. If you trust requests-mock, just look at the diff of this PR and if you do not see anything nefarious pull the specific commit you want (not the branch or a tag as it would be fairly easy for me to change the code you use).
I'll refrain from doing any further force push as it may suppress the commit you use.

Then, requests-mock can be reinstalled with:
```
pip install --force-reinstall requests-mock
```

jamielennox commented:
I apologise for being absent, it's been a crazy time for everyone but not a good excuse. 

I'm still shocked by how clean that implementation came out. It just seems to fit with the existing logic. Thanks for persevering.

I'm closing all the bugs that are open related to nesting in favour of this implementation. 

Two requests/questions if you have some time: 
   * I added some release notes, but do you mind writing up this in the docs, maybe an example. I suspect it'd be in just before https://github.com/jamielennox/requests-mock/blob/master/doc/source/mocker.rst#real-http-requests.
   * Any suggestions for what we do with `real_http`? It probably wasn't a great name to begin with but it seems you're more likely to use it for nesting than actual real http. Is there a name here that we can alias?

clslgrnc commented:
I'll probably be able to improve the doc in the next few weeks.
It just hit me that the behavior of nested mocks might be weird if one middle mock was stopped before the others (which is not possible with context managers).
I'll add a warning not to do that in the doc of `stop`.
It would be great if I could make it work though...

jamielennox commented:
Yea, that'd we a problem. Though the only way i can think to solve that (in the like 30 seconds since i've read that) would be maintain some seperate stack of mocks outside the mocker that you resolve up through. 

I did consider that once as a solution to nesting, however it's a decent size change for what is a very corner case. I'm ok with just a :warning: in the docs for now. 

Personally i'm still of the opinion you should avoid nesting mocks at all. I get that people need it for reasons, but if you start messing with the order you're doing something pretty wrong. 

I don't think there's much else in queue (except maybe #110 and i don't see that as urgent at the moment, and would need to reconcile it with this), so whenever you're good with that i can cut a release.

clslgrnc commented:
I updated #110 in #133 and updated the doc for both #110 and #116 in #134.
Let me know what you think.

As for a better name than `real_http`, I do not have any satisfying ideas.
